### DISCUSSION TITLE
PR testing "common pattern"

### DISCUSSION THREAD

**Discussion by PhillypHenning** [bitovi/bitops#397] [General]
As per our [internal Slack discussion](https://bitovi.slack.com/archives/C0129U3GL93/p1675437335475709)

I'd like to come to consensus on how a developer can test a bitops plugin within an github action workflow. 

**The challenge is;**
Currently BitOps branches do not create test images/tags, as a result, it is difficult to test changes to a plugin in a github workflow. For example, if a change is proposed for Terraform, and a PR is created, how could we test the PR? 

There isn't a straight forward answer today and there is no documentation to guide others.

**Comment by PhillypHenning** [comment]
What I did to accomplish some testing was; 

I created a `test` branch in the github action and update the `deploy.sh` script to point to a personal image of BitOps (I created and pushed to my account). 

Pro:
- Small update that is reflected immediately

Con:
- `deploy.sh` script will need to be returned to normal before being merged

**Comment by arm4b** [comment]
Yes, you definitely want to build a new image if the plugin installation steps changed (ansible/terraform version, new tools, etc).
But no, for rapid development it's less productive to rebuild the entire BitOps image if just the plugin deployment steps changed.

The alternative would be to specify the plugin dir (or repo w branch) to share with the BitOps container.

**Comment by arm4b** [comment]
Here is a comparison. If you have a WIP PR in the plugin repo (like https://github.com/bitops-plugins/terraform/pull/23), and push to that branch while iterating.

## 1. Workflow with BitOps Image rebuild
In this scenario for every new changed line of code in the plugin you have to rebuild the Docker image, wait for tools re-installation, docker build-> deploy. If it's not a one-time change (the majority of the work is more than that), then there's a lot of waiting time. There's also a need to open a PR in the BitOps repo that'll run some workflow for deploying and WIP plugin branch which is likely not relevant to the prod. So we have 3 places for context-switch: Plugin repo, BitOps repo + Docker images, GHA repo

## 2. Workflow with Plugin path sharing
In this scenario, you add the logic to git clone the plugin repo with the WIP branch, share it with the BitOps container, and re-run the workflow in the GHA repo. The downside is that additional logic needs to be added to clone stuff and [share dir](https://github.com/bitovi/github-actions-deploy-docker-to-ec2/blob/main/operations/_scripts/deploy/deploy.sh#L67) and then cleanup GHA deploy.sh once everything looks good.


You'll normally get faster results with the path-sharing approach rather than image rebuild.
But I think that for GHA work both approaches are not really great, especially considering we'll be doing more of this work.

**Comment by arm4b** [comment]
## 3. Random idea with `before` hooks
With https://bitops.sh/lifecycle/, what if we add the extra pull logic in the `before` script?
```sh
cd plugin-dir
git fetch
git checkout working-branch
```

This can potentially work as hooks are separated from the plugin run:
https://github.com/bitovi/bitops/blob/main/scripts/plugins/deploy_plugins.py#L247

How about trying that?
It would be nice if that works, as it relies on core BitOps functionality/design and definitely less hacky than others.

**Comment by arm4b** [comment]
Just tested with the `before` scripts, works perfectly:

Example [`bitops.before-deploy.d/1_switch-dev-branch.sh`](https://github.com/bitovi/github-actions-deploy-stackstorm/blob/7cc4108a852435ddccf4bc855d3476271a4cddc3/operations/deployment/ansible/bitops.before-deploy.d/1_switch-dev-branch.sh)
```
echo 'Testing the Ansible plugin WIP branch in dynamic GHA run'

cd ${BITOPS_INSTALLED_PLUGIN_DIR}
git fetch
git checkout fix_9
```

It's much easier than rebuilding a BitOps image or sharing the docker volumes as it allows to shortcut several operations in between.
